sklearn/preprocessing/target.py

```diff
@@ -0,0 +1,239 @@
+        self.regressor = regressor
+        self.transformer = transformer
+        self.func = func
+        self.inverse_func = inverse_func
+        self.check_inverse = check_inverse
+        # we probably need to change this ones we have tags
+        self._estimator_type = regressor._estimator_type
+
+    def _fit_transformer(self, y, sample_weight):
+        if (self.transformer is not None and
+                (self.func is not None or self.inverse_func is not None)):
+            raise ValueError("Both 'transformer' and functions 'func'/"
+                             "'inverse_func' cannot be set at the same time.")
+        elif self.transformer is not None:
+            self.transformer_ = clone(self.transformer)
+        else:
+            self.transformer_ = FunctionTransformer(
+                func=self.func, inverse_func=self.inverse_func, validate=False)
+        fit_parameters = signature(self.transformer_.fit).parameters
+        if "sample_weight" in fit_parameters:
+            self.transformer_.fit(y, sample_weight=sample_weight)
+        else:
+            self.transformer_.fit(y)
+        self.transformer_.fit(y)
+        if self.check_inverse:
+            n_subsample = min(1000, y.shape[0])
+            subsample_idx = np.random.choice(range(y.shape[0]),
+                                             size=n_subsample, replace=False)
+            if not np.allclose(
+                    y[subsample_idx],
+                    self.transformer_.inverse_transform(
+                        self.transformer_.transform(y[subsample_idx]))):
+                raise ValueError("The provided functions or transformer are"
+                                 " not strictly inverse of each other.")
+
+    def fit(self, X, y, sample_weight=None):
+        """Fit the model according to the given training data.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape (n_samples, n_features)
+            Training vector, where n_samples is the number of samples and
+            n_features is the number of features.
+
+        y : array-like, shape (n_samples,)
+            Target vector relative to X.
+
+        sample_weight : array-like, shape (n_samples,) optional
+            Array of weights that are assigned to individual samples.
+            If not provided, then each sample is given unit weight.
+
+        Returns
+        -------
+        self : object
+            Returns self.
+        """
+        y = check_array(y, ensure_2d=False)
+        # memorize if y should be a multi-output
+        self.y_ndim_ = y.ndim
+        if y.ndim == 1 and self.func is None:
+            y_2d = y.reshape(-1, 1)
+        else:
+            y_2d = y
+        self._fit_transformer(y_2d, sample_weight)
+        self.regressor_ = clone(self.regressor)
+        if sample_weight is not None:
+            self.regressor_.fit(X, self.transformer_.transform(y_2d),
```

Comment:
I mean that we should really be using fit_transform to produce the downstream y here